### DISCUSSION TITLE
Adding other optimization algorithm to the Dafoam

### DISCUSSION THREAD

**Discussion by qwefhajk** [mdolab/dafoam#138] [General]
I want to run the topology optimization case but I do not have the acess to SNOPT. Therefore, I wander how to add other optimization algorithm like the MMA/GCMMA in the pyopt to the Dafoam? I am just getting started on Dafoam by using the Docker image form. Should I compile it form source instead or is there a way to add pyOpt into the Dafoam docker image.

**Comment by friedenhe** [comment]
Unfortunately, the MMA/CGMMA optimizers were not added in pyOptSparse. You may need to create an issue on the pyOptSparse repo and request this feature. Once added, you need to compile everything from scratch, or you can compile the new pyOptSparse in the DAFoam Docker image and save it as a new image. Note that our framework does not support pyOpt. We support only pyOptSparse.

**Reply by qwefhajk** [reply]
Thanks for your reply! May I ask ask another quetions? How dose the variable 'alphaPorosity' be sent to the primal equation and which files are involved ? I am trying to figure out the flowchart of the topology optimization code.

**Reply by friedenhe** [reply]
The alphaPorosity field is load from [here](https://github.com/mdolab/dafoam/blob/dc1f08f6d114e2dceae0e90dd72ce38c9039bdba/src/adjoint/DASolver/DASimpleTFoam/createFieldsSimpleT.H#L42). It is then used in the U equation from [here](https://github.com/mdolab/dafoam/blob/dc1f08f6d114e2dceae0e90dd72ce38c9039bdba/src/adjoint/DASolver/DASimpleTFoam/UEqnSimpleT.H#L5). The alphaPorosity field is assigned directly by the optimizer through this [function](https://github.com/mdolab/dafoam/blob/dc1f08f6d114e2dceae0e90dd72ce38c9039bdba/src/adjoint/DASolver/DASolver.C#L4404).

**Reply by qwefhajk** [reply]
Thanks for your detailed information again !! I intend to implement my version of MMA algorithm. Therefore I have a question,  directly using the function "optFuncs.calcObjFuncValues" and "optFuncs.calcObjFuncSens" under the mpirun mode may lead to false information about the Objective and sensitivity, am I right? Can you shed a light on the correct processment to obtain the Objective and senstivity regarding these two function? I notice that in the original code this is processed by the pyoptsparse.Optimization module.

**Reply by friedenhe** [reply]
Can you be more specific about: "directly using the function .. may lead to false information."? Maybe a log file can provide more details.

**Reply by qwefhajk** [reply]
For instance, in the code segment of ["if args.task == "opt""](https://github.com/DAFoam/tutorials/blob/master/pitzDaily/runScript.py) in line 170. I choose to bypass the optimization module and simlpy implement the code
`DASolver.runColoring()`
`x = alphaPorosity0.copy()`
`    for idxI, v in enumerate(x):`
`        DASolver.setFieldValue4GlobalCellI(b"alphaPorosity", v, idxI)`
`        DASolver.updateBoundaryConditions(b"alphaPorosity", b"scalar")`
`    xdv = {'alphaPorosity':x}`
`objFun = optFuncs.calcObjFuncValues(xdv)`

and I found out that, the results "objFun" are different for implementing "mpirun np -4" or direclty running it without mpi.
The results are 0.28228330, 0.28227995, 0.28227893, 0.28229530 respectively for without MPI, np-2, np-3 and np-4.
I do not know which part I have missed to lead to this problem.

**Reply by friedenhe** [reply]
The way you call calcObjFuncValues is OK. But, usually, if we want to manually modify or assign values for design variables, we do something like this:

designVar = DVGeo.getValues()
designVar["alphaPorosity"][0] = some value
designVar["alphaPorosity"][1] = some other value
...
objFun = optFuncs.calcObjFuncValues(designVar)

For your question, what is alphaPorosity0? Is it uniform or all zeros? If it is uniform the objFun should match between your implementation and the default task==opt one. Also note that, if alphaPorosity0 is not uniform, the objFunc will be different if you use different numbers of CPU cores. This is because, in parallel, OpenFOAM rearranges the indexing of cells. So the order in alphaPorosity0 will be different. For example, if in serial, alphaPorosity = {cell0, cell1, cell2, cell3}, in parallel, proc may own {cell0, cell3} and proc2 may own {cell1, cell2}, so the alphaPorosity variable in parallel will be alphaPorosity = {cell0, cell3, cell1, cell2}. You may also want to refer to the discussion in this [post](https://github.com/mdolab/dafoam/discussions/102)

**Reply by qwefhajk** [reply]
In fact, the alphaPorosity0 is all zeros as in the original code and I still get different results. And is there some way to secure the proper index of the cells for alphaPorosity0 in parallel mode?

**Reply by friedenhe** [reply]
It is hard to tell where the issue came from. Can you send me the case to reproduce the issue?

**Reply by qwefhajk** [reply]
Thanks a lot for taking time to look at my case. And I notice that it gets some warnings like complexWarning for function "setFieldValue4GlobalCellI", only in the MPI mode. The code is listed as below.
[runScript322.txt](https://github.com/mdolab/dafoam/files/6652551/runScript322.txt)

**Reply by friedenhe** [reply]
The reason the simulation results do not 100% match in serial and in parallel is that the flow is not tightly converged. If you set  `"primalMinResTol": 1.0e-8`, you should see more digits match.

**Reply by qwefhajk** [reply]
Thanks you for the tips. I will look into it with further tests. Really appreciate your help!

**Reply by hosseinnegahban** [reply]
> Unfortunately, the MMA/CGMMA optimizers were not added in pyOptSparse. You may need to create an issue on the pyOptSparse repo and request this feature. Once added, you need to compile everything from scratch, or you can compile the new pyOptSparse in the DAFoam Docker image and save it as a new image. Note that our framework does not support pyOpt. We support only pyOptSparse.

Could you please be more specific on how to  compile a new pyOptSparse in the DAFoam Docker image?

**Reply by friedenhe** [reply]
https://dafoam.github.io/mydoc_installation_docker.html

**Reply by Ananth-Narayan-IITM** [reply]
@friedenhe, I was infact wanted MMA just to ensure how well it behaves than SNOPT and managed to make this work with pyOptSparse. I did install new version of pyOptSparse and PyOpt (which has MMA). Please refer to instructions which worked for me.

Edit1: 
I will try to release few examples and a solver for cases using MMA

- ## Installing ParOpt (for MMA)- Topology Optimization in Docker
	- ### ParOpt
		- Clone the repository and extract
		  ```bash
		  cd ~/dafoam/repos
		  wget https://github.com/smdogroup/paropt/archive/refs/tags/v2.1.5.tar.gz
		  tar -xvf v2.1.5.tar.gz
		  cd paropt-2.1.5
		  cp Makefile.in.info Makefile.in
		  export METIS_DIR="/home/dafoamuser/dafoam/repos/tacs/extern/metis"
		  ```
		- Make sure that `Makefile.in` file is edited such that `PAROPT_DIR=/home/dafoamuser/dafoam/repos/paropt-2.1.5` (or other location pointing paropt)
		- Make sure packages and modules are installed as given in `INSTALL.txt`, 
		  ```bash
		  sudo apt-get install libblas-dev -y
		  sudo apt-get install liblapack-dev -y
		  sudo apt-get install libopenmpi-dev
		  sudo apt-get install openmpi-bin
		  pip3 install --upgrade pip
		  pip3 install numpy
		  pip3 install Cython
		  pip3 install mpi4py
		  pip3 install matplotlib
		  ```
		  and export `PYTHONPATH` and `PAROPT_DIR` as given  in `INSTALL.txt`
		- Now compile them,
		  ```bash
		  make && pip install -e .[all]
		  ```
		- Verify the ParOpt installation in Python as
		  ```python
		  import paropt 
		  print(paropt.__file__)
		  ```
		  Expected output is `/home/dafoamuser/dafoam/repos/paropt-2.1.5/paropt/__init__.py`  
	- ### pyOptSparse
		- Uninstall current `pyOptSparse v2.10.1`,
		  ```bash
		  pip uninstall pyoptsparse
		  ```
		- Verify if uninstall is success by `python -c "import pyoptsparse"`
		- Clone the `pyOptSparse v2.15.0`, which has Python Compatibility of `>=3.9` thus suitable for the DAFoam docker (having `Python 3.10.8`).
		  ```bash
		  cd ~/dafoam/repos
		  git clone https://github.com/mdolab/pyoptsparse.git
		  cd pyoptsparse
		  git checkout v2.15.0
		  pip install .
		  ```
	- ### Compatibility check between `pyOptSparse` and `ParOpt` versions
		- It was noticed that from `paropt/paropt_pyoptsparse.py` had issue with `ImportError: cannot import name 'Error'`. However if you are using latest version of `ParOpt` make sure to check the next step and act accordingly
		- From `paropt/paropt_pyoptsparse.py` replace line 13 from `from pyoptsparse.pyOpt_error import Error` to
		  ```python
		  try: 
		  	from pyoptsparse.pyOpt_error import Error 
		  except ImportError: 
		      pass
		  ```
		- Make sure to replace if this line exists. I noticed that in recent version this has been removed
		- To verify the integration, launch Python
		  ```python
		  from pyoptsparse import OPT 
		  opt = OPT("ParOpt") 
		  print(opt) 
		  ```
		  which outputs similar to `<paropt.paropt_pyoptsparse.ParOptSparse object at ...>`  
	- Upon verification, make to compile the DAFoam code using `./Allmake` under respective directory
	- Further verify this by running the cases by modifying `runScript.py` as
	  ```python
	  parser.add_argument( "-optimizer", type=str, default="ParOpt")
	  # Under args.optimizer
	  elif args.optimizer == "ParOpt":
	      prob.driver.opt_settings = {
	          "algorithm": "mma",
	      }
	  ```

**Reply by friedenhe** [reply]
@Ananth-Narayan-IITM  Thanks very much for sharing the installation guidelines. We will incorporate this into our new DAFoam version and update the website.
